src/editor/helpers.ts

```diff
@@ -1,71 +1,226 @@
-import { Editor, Transforms, Range, Text } from "slate";
+import { Editor, Transforms, Location, Range } from "slate";
 import isHotkey from "is-hotkey";
 
-import { Mark, SluglineElement } from "./types";
+import {
+  Mark,
+  InlineElement,
+  SluglineElement,
+  InlineVoidElement,
+  BlockElementType,
+  InlineElementType,
+} from "./types";
+import { HistoryEditor } from "slate-history";
 
-export function isMarkActive(editor: Editor, mark: Mark): boolean {
+export const isMarkActive = (editor: Editor, mark: Mark): boolean => {
   const marks = Editor.marks(editor);
   return marks !== null && marks[mark] === true;
-}
+};
 
-export function toggleMark(editor: Editor, mark: Mark): void {
+export const toggleMark = (editor: Editor, mark: Mark): void => {
   const active = isMarkActive(editor, mark);
   if (active) {
     Editor.removeMark(editor, mark);
   } else {
     Editor.addMark(editor, mark, true);
   }
-}
-
-export function createInline(editor: Editor, inline: SluglineElement): void {
-  Transforms.insertNodes(editor, inline);
-}
-
-export function wrapInline(editor: Editor, inline: SluglineElement): void {
-  // remove any children since we're wrapping an existing node
-  const topLevel = {
-    ...inline,
-    children: []
-  };
-  Transforms.wrapNodes(editor, topLevel, {
-    split: true
+};
+
+const clearAllStress = (editor: Editor) => {
+  Editor.removeMark(editor, Mark.Stress1);
+  Editor.removeMark(editor, Mark.Stress2);
+};
+
+const clearAllEmph = (editor: Editor) => {
+  Editor.removeMark(editor, Mark.Emph1);
+  Editor.removeMark(editor, Mark.Emph2);
+  Editor.removeMark(editor, Mark.Emph3);
+  Editor.removeMark(editor, Mark.Emph4);
+};
+
+export const increaseStress = (editor: Editor) => {
+  clearAllEmph(editor);
+
+  if (isMarkActive(editor, Mark.Stress1)) {
+    clearAllStress(editor);
+    editor.addMark(Mark.Stress2, true);
+  } else if (isMarkActive(editor, Mark.Stress2)) {
+    clearAllStress(editor);
+  } else {
+    // no marks at all, add the first level
+    editor.addMark(Mark.Stress1, true);
+  }
+};
+
+export const increaseEmph = (editor: Editor) => {
+  clearAllStress(editor);
+
+  if (isMarkActive(editor, Mark.Emph1)) {
+    clearAllEmph(editor);
+    editor.addMark(Mark.Emph2, true);
+  } else if (isMarkActive(editor, Mark.Emph2)) {
+    clearAllEmph(editor);
+    editor.addMark(Mark.Emph3, true);
+  } else if (isMarkActive(editor, Mark.Emph3)) {
+    clearAllEmph(editor);
+    editor.addMark(Mark.Emph4, true);
+  } else if (isMarkActive(editor, Mark.Emph4)) {
+    clearAllEmph(editor);
+  } else {
+    editor.addMark(Mark.Emph1, true);
+  }
+};
+
+/**
+ * Returns true if `iterable` has no elements.
+ * @param iterable The iterable to check
+ */
+const isIterableEmpty = (iterable: Iterable<any>) => {
```

Comment:
Suggestion: Add a check if the iterable is an array so you can do a simple length comparison, something like

```suggestion
const isIterableEmpty = (iterable: Iterable<any>) => {
  if (Array.isArray(iterable)) return iterable.length == 0; // or `if ("length" in iterable)`, though that may be slower
```

Reply:
Actually ignore this as this will def be slower than loop unrolling.